### PR TITLE
Create `xxxContext` functions for ones which are missing a caller context.

### PR DESCRIPTION
**Pull request by JoshVanL**
PR adds `xxxContext` variations of functions that are currently using a `context.Background()`. This enables callers to pass their own context.

For backwards compat, no existing functions have changed their signature and behave the same as before- they are now all a wrapper for their `xxxContext` counterparts.

### CODE REVIEW

**Comment by yaron2** [comment]
@daixiang0 please help review.

File: actor/actor.go [hunk 1]
```diff
@@ -78,25 +81,40 @@ func (b *ServerImplBase) SetID(id string) {
...
 	}
 	return nil
 }
 
+//nolint:interfacebloat
 type StateManager interface {
 	// Add is to add new state store with @stateName and @value
 	Add(stateName string, value interface{}) error
+	// Add is the same as Add, but with context.
+	AddContext(ctx context.Context, stateName string, value interface{}) error
```
Comment (daixiang0) [inline actor/actor.go:99]: I would like `AddWithContext()` to include `Add()`, or make `AddContext()`  only to do one thing -- add context.

Comment (artursouza) [reply actor/actor.go:99]: Would it make more sense to deprecate this interface and create a new one? The existing one would still be supported but fallback to calling the new implementation where context is always present.

Comment (daixiang0) [reply actor/actor.go:99]: I agree.

@JoshVanL please update as @artursouza said.

File: actor/actor.go [hunk 2]
```diff
@@ -39,64 +36,176 @@ type Server interface {
...
 	id           string
+	lock         sync.RWMutex
 }
 
+// Deprecated: Use ServerImplBaseCtx instead.
 func (b *ServerImplBase) SetStateManager(stateManager StateManager) {
+	b.ctx.lock.Lock()
+	defer b.ctx.lock.Unlock()
 	b.stateManager = stateManager
+	b.ctx.stateManager = stateManager.WithContext()
```
Comment (artursouza) [inline actor/actor.go:95]: Does this create a race condition where two contexts can be set while the invocation did not take place yet? Is there a lock somewhere surrounding WithContext() and SaveState()?

The lock here will avoid another race condition instead.

File: actor/actor.go [hunk 3]
```diff
@@ -39,64 +36,176 @@ type Server interface {
...
 	Contains(stateName string) (bool, error)
 	// Save is to saves the state cache of this actor instance to state store component by calling api of daprd.
 	Save() error
-	// Flush is called by stateManager after Save
+	// Flush is called by StateManager after Save
 	Flush()
+
+	// Returns a new StateManagerContext with the same state as this StateManager
+	// but uses context.
+	WithContext() StateManagerContext
```
Comment (artursouza) [inline actor/actor.go:193]: Same question about race condition.

**Review by artursouza — CHANGES_REQUESTED** [review]

File: actor/actor.go [hunk 4]
```diff
@@ -39,64 +36,176 @@ type Server interface {
...
 	}
+
 	return nil
 }
 
+// Deprecated: StateManager is deprecated in favour of StateManagerContext.
 type StateManager interface {
 	// Add is to add new state store with @stateName and @value
-	Add(stateName string, value interface{}) error
+	Add(stateName string, value any) error
```
[the remark this answers is not in the document]
Comment (JoshVanL) [reply actor/actor.go:177]: No, `any` is a straight type alias to `interface{}`.

File: actor/actor.go (hunk 2 above)
Comment (JoshVanL) [reply actor/actor.go:95]: Thanks, there can be. I'll add another lock to `ServerImpBase`.

**Comment by JoshVanL** [comment]
Thanks @artursouza, PTAL

File: actor/actor.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (daixiang0) [reply actor/actor.go:184]: let's keep it as before.

[the remark this answers is not in the document]
Comment (JoshVanL) [reply actor/actor.go:184]: @daixiang0 any particular reason for this? The `any` keyword seems cleaner, is far more preferred in upstream go.

```bash
$ grep -inr /nix/store/ngvd1babcfc84a2rsdsvbrlmc6w3pg8m-go-1.20.1/share/go/src --exclude-dir=internal -e "any" | wc -l
8564
```

```bash
$ grep -inr /nix/store/ngvd1babcfc84a2rsdsvbrlmc6w3pg8m-go-1.20.1/share/go/src --exclude-dir=internal -e "interface{}" | wc -l
447
```

**Review by artursouza — APPROVED** [review]

**Pull request merged**